README.md
EnlighterJS Plugin for Gutenberg Editor
=========================================

![Gutenberg Editor Integration](assets/enlighterjs_gutenberg_test.png)

Current Limitations
----------------------------------------------

* Legacy EnlighterJS code cannot be transformed to blocks by the "Transform to Blocks" utility of the classic editor.  
**Caused By:** [gutenberg/8648](https://github.com/WordPress/gutenberg/issues/8648)
* Newly created codeblocks (via Gutenberg) mess up the HTML with unused attributes because of a validation issue.  
**Caused By:** [gutenberg/8532](https://github.com/WordPress/gutenberg/issues/8532) / [gutenberg/7604](https://github.com/WordPress/gutenberg/issues/7604)

Usage
----------------------------------------------

As integral part of the Enlighter WordPress Plugin (future)

Development
----------------------------------------------

1. Create the plugin directory `wp-content/plugins/enlighter-gutenberg`
2. Copy `enlighter-gutenberg.php` and `dist/` into this directory
3. Activate the plugin

**Required Directory Structure**

```
wp-content/
   | - plugins/
   |      | -enlighter-gutenberg/
   |      |       |- enlighter-gutenberg.php
   |      |       |- dist/
   |      |       |      |- enlighterjs.gutenberg.min.css 
   |      |       |      |- enlighterjs.gutenberg.min.js
```

License
----------------------------------------------

EnlighterJS.Gutenberg is OpenSource and licensed under the Terms of [Mozilla Public License 2.0](https://opensource.org/licenses/MPL-2.0). You're welcome to [contribute](docs/CONTRIBUTING.md)


CHANGES.md
## Branch 1.x ##

### 1.2.0 ###

* Added: Gutenberg left-align-indentation formatting button
* Added: block transformation from SyntaxHighlighter Evolved
* Added: selected language as button text to the block toolbar
* Bugfix: Gutenberg inline editing not working on Safari due to required `rangeCount` call - thanks to [phruse on GitHub](https://github.com/EnlighterJS/Plugin.Gutenberg/issues/36)
* Bugfix: Gutenberg inline popover window content is collapsed in recent Gutenberg releases
* Bugfix: Gutenberg editor deprecation warnings related to Toolbar elements

### 1.1.1 ###

* Bugfix: Gutenberg `escapeEditableHTML` throws an error in case the block is empty (attributes.content returns `null`)

### 1.1.0 ###

* Added: inline highlighting support (RichText format)
* Added: border around Enlighter codeblock (default theme)
* Changed: updated buildsystem to babel `v7` and gulp `v4`
* Bugfix: escape sequence were transformed into their characters in recent Gutenberg releases

### v1.0.0 ###

* Added: custom css class attributes to generated pre blocks

## PRELIMINARY RELEASE ##

### v0.4.0 ###

* Added: automatically transforms legacy Enlighter codeblocks (Classic Editor) to Gutenberg blocks in case the "Convert to Blocks" function is triggered
* Changed: direction `ltr` is enforced within the editing area - thanks to [nekofar on GitHub](https://github.com/EnlighterJS/Plugin.Gutenberg/pull/9)

src/blocks/codeblock.js
// ----------------------------------------------------------------------
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
// --
// Copyright 2018 Andi Dittrich <https://andidittrich.de>
// ----------------------------------------------------------------------

import {Editor, Components, Element, Blocks} from '../wp';
import _TextEditor from '../components/TextEditor';
import _LanguageMenu from '../components/blockcontrols/LanguageMenu';
import _LanguageSelect from '../components/inspectorcontrols/LanguageSelect';
import _ThemeSelect from '../components/inspectorcontrols/ThemeSelect';
import {getLanguageLabel} from '../languages';

// Standard Codeblock
export default {

    // metadata
    title: 'Enlighter Sourcecode',
    description: 'Syntax highlighting with EnlighterJS.',
    icon: 'editor-code',
    category: 'formatting',
    keywords: ['code', 'sourcecode', 'enlighter'],

    // internal block state
    // @see https://wordpress.org/gutenberg/handbook/block-api/attributes/
    attributes: {
        // extract sourcecode from saved (html) content
        content: {
            type: 'string',
            selector: 'pre.EnlighterJSRAW',
            source: 'text'
        },

        // Extract EnlighterJS related attributes
        // use type attribute instead of query for easier handling
        language: {
            type: 'attribute',
            attribute: 'data-enlighter-language', 
            default: 'generic'
        },

        theme: {
            type: 'attribute',
            attribute: 'data-enlighter-theme', 
            default: ''
        },

        highlight: {
            type: 'attribute',
            attribute: 'data-enlighter-highlight', 
            default: ''
        },

        linenumbers: {
            type: 'attribute',
            attribute: 'data-enlighter-linenumbers', 
            default: ''
        },

        lineoffset: {
            type: 'attribute',
            attribute: 'data-enlighter-lineoffset', 
            default: ''
        },

        title: {
            type: 'attribute',
            attribute: 'data-enlighter-title', 
            default: ''
        }
    },

    // extended block support features
    // @see https://wordpress.org/gutenberg/handbook/block-api/#supports-optional
    supports: {
        // no custom classes
        customClassName: false,

        // remove auto generated wrapper classname
        className: false,

        // disable html edit mode
        html: false
    },

   // block element transformation/converting
    transforms: {
        from: [

            // allow transform from RAW DOM <pre><code> (legacy syntax highlighting)
            // higher priority then 'core/preformatted'
            // automatically applied when transforming from legacy to blocks
            {
                type: 'raw',
                priority: 4,
                isMatch: function(node){
                    return  node.nodeName === 'PRE' &&
                            node.children.length === 1 &&
                            node.firstChild.nodeName === 'CODE'
                },
                transform: function (node){
                    // use inner text as content
                    return Blocks.createBlock('enlighter/codeblock', {content: node.textContent});
                },
            },

            // allow transform from standard EnlighterJS code to blocks
            // higher priority then 'core/preformatted'
            // automatically applied when transforming from legacy to blocks
            // @BUG https://github.com/WordPress/gutenberg/issues/8648
            // @TODO re-enable transform and use attribute matching
            /*
            {
                type: 'raw',
                priority: 4,
                isMatch: function(node){
                    return  false;
                },
                transform: function (node){
                    // use inner text as content
                    return _wp.blocks.createBlock('enlighter/codeblock', {content: node.textContent});
                },
            },
            */

            // allow transform from core/code block
            // allow transform from core/preformatted block
            // allow transform from core/paragraph block
            {
                type: 'block',
                blocks: ['core/code', 'core/preformatted', 'core/paragraph'],
                transform: function ({content}) {
                    return Blocks.createBlock('enlighter/codeblock', {content});
                },
            }
        ],

        to: [
            // allow transform to core/code block
            {
                type: 'block',
                blocks: ['core/code'],
                transform: function ({content}) {
                    return Blocks.createBlock('core/code', {content});
                },
            },

            // allow transform to core/preformatted block
            {
                type: 'block',
                blocks: ['core/preformatted'],
                transform: function ({content}) {
                    return Blocks.createBlock('core/preformatted', {content});
                },
            }
        ]
    },

    // The "edit" property must be a valid function.
    edit: function ui({attributes, setAttributes}){

        // use standard Gutenberg PlainText View with custom styles
        return <Element.Fragment>

            <Editor.BlockControls>
                <Components.Toolbar>
                    <_LanguageMenu value={attributes.language} onChange={(language) => setAttributes({language})}/>
                </Components.Toolbar>
            </Editor.BlockControls>

            <div className="enlighter-block-wrapper">
                <div className="enlighter-header">
                    <div className="enlighter-title">{getLanguageLabel(attributes.language)}</div>
                </div>
                <_TextEditor
                    value={ attributes.content }
                    onChange={(content) => setAttributes({content})}
                    placeholder="Insert Sourcecode.."
                    aria-label="Code"
                />
                <div className="enlighter-footer">
                    <div className="enlighter-footer-label"><strong>EnlighterJS</strong> Syntax Highlighter</div>
                </div>
            </div>

            <Editor.InspectorControls>
                <Components.PanelBody title="Appearance">
                    <p>Override the global plugin settings of the current block.</p>
                    <_LanguageSelect value={attributes.language} onChange={(language) => setAttributes({language})} />
                    <_ThemeSelect value={attributes.theme} onChange={(theme) => setAttributes({theme})} />
                    <Components.ToggleControl
                        label="Linenumbers"
                        checked={ attributes.linenumbers }
                        onChange={(linenumbers) => setAttributes({linenumbers})}
                    />
                </Components.PanelBody>
            </Editor.InspectorControls>

        </Element.Fragment>
        ;
    },

    // render element as html
    // @BUG https://github.com/WordPress/gutenberg/issues/8532
    // @TODO use custom validator 
    save: function render({attributes}){
        // add enlighterjs related attributes
        return <pre 
            className="EnlighterJSRAW"
            data-enlighter-language={attributes.language}
            data-enlighter-theme={attributes.theme}
            data-enlighter-highlight={attributes.highlight}
            data-enlighter-linenumbers={attributes.linenumbers}
            data-enlighter-lineoffset={attributes.lineoffset}
            data-enlighter-title={attributes.title}
            >
            {attributes.content}
            </pre>;
    }
};